README.md

```diff
@@ -511,6 +534,721 @@ will essentially perform a
+    in
+    Xt.commit { tx } ()
+val n_elems : int = 2
+```
+
+### Race to cooperate
+
+Sometimes it is necessary to perform slower transactions that access many shared
+memory locations or need to perform expensive computations during the
+transaction. As mentioned previously, such transactions are more likely to
+suffer from interference or even starvation as other transactions race to make
+conflicting mutations to shared memory locations. To avoid such problems, it is
+often possible to split the transaction into two:
+
+1. A quick transaction that adversarially races against others.
+2. A slower transaction that others will then cooperate to complete.
+
+#### A three-stack lock-free queue
+
+Recall the [two-stack queue](#a-transactional-lock-free-queue) discussed
+earlier. The problem is that the `try_dequeue` operation `rev`erses the `back`
+of the queue and that can be relatively expensive. One way to avoid that problem
+is to introduce a third "middle" stack, or shared memory location, to the queue
+and quickly move the back to the middle stack.
+
+First we redefine the `queue` type to include a `middle` location:
+
+```ocaml
+# type 'a queue = {
+    back : 'a list Loc.t;
+    middle : 'a list Loc.t;
+    front : 'a list Loc.t;
+  }
+type 'a queue = {
+  back : 'a list Loc.t;
+  middle : 'a list Loc.t;
+  front : 'a list Loc.t;
+}
+```
+
+And adjust the `queue` constructor function accordingly:
+
+```ocaml
+# let queue () =
+    let back = Loc.make []
+    and middle = Loc.make []
+    and front = Loc.make [] in
+    { back; middle; front }
+val queue : unit -> 'a queue = <fun>
+```
+
+The `enqueue` operation remains essentially the same:
+
+```ocaml
+# let enqueue ~xt queue elem =
+    Xt.modify ~xt queue.back @@ List.cons elem
+val enqueue : xt:'a Xt.t -> 'b queue -> 'b -> unit = <fun>
+```
+
+For the quick transaction we introduce a helper function:
+
+```ocaml
+# let back_to_middle queue =
+    let tx ~xt =
+      match Xt.exchange ~xt queue.back [] with
+      | [] -> raise Not_found
+      | xs ->
+        if Xt.exchange ~xt queue.middle xs != [] then
+          raise Not_found
+    in
+    try Xt.commit { tx } with Not_found -> ()
+val back_to_middle : 'a queue -> unit = <fun>
+```
+
+Note that the above uses the `Not_found` exception to abort the transaction in
+case the `back` is empty or the `middle` already contains elements.
+
+The `dequeue` operation then runs the quick transaction to move elements from
+the `back` to the `middle` before examining the `middle`:
+
+```ocaml
+# let dequeue ~xt queue =
+    match Xt.update ~xt queue.front tl_safe with
+    | x :: _ -> Some x
+    | [] ->
+      back_to_middle queue;
+      match Xt.exchange ~xt queue.middle [] |> List.rev with
+      | x :: xs ->
+        Xt.set ~xt queue.front xs;
+        Some x
+      | [] ->
+        match Xt.exchange ~xt queue.back [] |> List.rev with
+        | x :: xs ->
+          Xt.set ~xt queue.front xs;
+          Some x
+        | [] -> None
```

Comment:
I think more comments are needed here. Committing  a transaction inside a transaction feels pretty unnatural to me. I will try to explain why.

First, there is this idea that a transaction is something that is going to be reversed if it failed. But the internal committed transaction is a side effect that is not going to be reversed. 

Second, it seems to kind of defeat one of most interesting part of transactions : writing code in a sequential way, without considering what other domains are doing. It is very clear here : `back_to_middle` is called, and ones could expect `queue.middle` to not be empty since we just move `queue.front` in it. But, it could actually be empty because this operation is not part of the transaction. 

Also, this creates misleading code as this one: 
```ocaml
let test () =
  let switch_3_to_2 a =
    let tx ~xt =
      if Xt.get ~xt a = 3 then Xt.set ~xt a 2 in
    Xt.commit {tx}
  in
  let tx ~xt x  =
    Xt.set ~xt x 3;
    switch_3_to_2 x;
  in
  let v = Loc.make 0 in
  Xt.commit {tx = tx v};
  Loc.get v
````
that returns `3`. 


May be a useful comment for #30.

Reply:
> I think more comments are needed here.

Yes, I think that is fair observation.

The algorithm design technique described in this section is definitely more advanced than the fairly basic tips given in other sections.

> Committing a transaction inside a transaction [...]

It is actually a common pattern, or algorithm design technique, used in lock-free algorithms to _perform an atomic operation while preparing to perform another atomic operation_.

The emphasized part of the previous sentence can be argued to be a rephrase of what you said. A transaction (function) is something that can be run to prepare a specification of operations (the transaction log) to be attempted atomically (by the k-CAS-n-CMP algorithm).

(I wanted to highlight that part of what you said to draw the analogy (not trying to e.g. quote you out of context).)

> ...] feels pretty unnatural to me. I will try to explain why.

Yes, this is definitely an advanced technique, much like the whole of designing lock-free algorithms.

> First, there is this idea that a transaction is something that is going to be
> reversed if it failed.

Hmm... I would not say that a "transaction is [...] reversed if it failed", because that could give false impressions &mdash; Transactions aren't reversible and the construction of a transaction log does not (logically) mutate shared memory locations.

As a side note, I used "logically" above, because due to the underlying k-CAS-n-CMP algorithm, reading the content of a shared memory location may require the reader to help a potential writer to determine the content of the location being read. But I digress.

What the documentation says, for example, is that transactions are "attempted" and that they are "retried" in case of failure and the transaction logs are "constructed" by "record"ing operations.

> But the internal committed transaction is a side effect that is not going to
> be reversed.

Yes, that is basically analogous to the common technique of "helping" in lock-free algorithms.

In the Michael-Scott queue, for example, operations on the queue help each other to keep the tail updated. Updates of the tail do not change the logical state of the queue, but work towards making operations efficient and lock-free.

Similarly moving a non-empty back to the previously empty middle is a step that does not change the logical state of the queue, but is something that works towards that goal and multiple consumers can work together to make that happen.

The difference is that with **kcas** such helping operations are not limited to 1-CAS or single-word compare-and-set.

> Second, it seems to kind of defeat one of most interesting part of
> transactions : writing code in a sequential way, without considering what
> other domains are doing.

Well, yes and no.

- _Yes_ &mdash; When you use these kinds of lock-free algorithm design techniques, you make it more difficult to reason about the internal algorithmic details of transactions.

- _No_ &mdash; If you use the techniques correctly (making sure that the helping side-effects do not change the logical results) you can implement more fine-grained lock-free algorithms that can then be used compositionally without having to understand the internal details.

---

Thanks for raising these questions!

The technique obviously needed more fleshing out than what I had written before. As I've argued, the technique itself is not out of the ordinary &mdash; I would argue that it is based on common lock-free algorithm design techniques. It requires a good understanding of concurrency issues and the transaction mechanism and opens the possibility of writing more fine-grained and better performing lock-free algorithms. This is similar to what Reagents offer, but using the transaction mechanism. Furthermore, in a language like OCaml it would be difficult to completely prevent it, because OCaml makes it easy to perform side-effects anywhere.

I've added more elaboration on the topic to the documentation.

Reply:
Thanks for the changes you made, it answers well my concerns !